databroker/databroker.py

```diff
@@ -11,96 +11,144 @@
 import metadatastore.commands as mc
 import filestore.api as fs
 import logging
+import numbers
+
+
+try:
+    from functools import singledispatch
+except ImportError:
+    try:
+        # We are running on Python 2.6, 2.7, or 3.3
+        from singledispatch import singledispatch
+    except ImportError:
+        raise ImportError(
+            "Please install singledispatch from PyPI"
+            "\n\n   pip install singledispatch"
+            "\n\nThen run your program again."
+        )
+try:
+    from collections.abc import MutableSequence
+except ImportError:
+    # This will error on python < 3.3
+    from collections import MutableSequence
 
 
 logger = logging.getLogger(__name__)
 TZ = str(tzlocal.get_localzone())
 
 
+@singledispatch
+def search(key):
+    logger.info('Using default search for key = %s' % key)
+    raise ValueError("Must give an integer scan ID like [6], a slice "
+                     "into past scans like [-5], [-5:], or [-5:-9:2], "
+                     "a list like [1, 7, 13], a (partial) uid "
+                     "like ['a23jslk'] or a full uid like "
+                     "['f26efc1d-8263-46c8-a560-7bf73d2786e1'].")
+
+
+@search.register(slice)
+def _(key):
+    # Interpret key as a slice into previous scans.
+    logger.info('Interpreting key = %s as a slice' % key)
+    if key.start is not None and key.start > -1:
+        raise ValueError("slice.start must be negative. You gave me "
+                         "key=%s The offending part is key.start=%s"
+                         % (key, key.start))
+    if key.stop is not None and key.stop > 0:
+        raise ValueError("slice.stop must be <= 0. You gave me key=%s. "
+                         "The offending part is key.stop = %s"
+                         % (key, key.stop))
+    if key.stop is not None:
+        stop = -key.stop
+    else:
+        stop = None
+    if key.start is None:
+        raise ValueError("slice.start cannot be None because we do not "
+                         "support slicing infinitely into the past; "
+                         "the size of the result is non-deterministic "
+                         "and could become too large.")
+    start = -key.start
+    result = list(find_last(start))[stop::key.step]
+    header = [Header.from_run_start(h) for h in result]
+    return header
+
+
+@search.register(numbers.Integral)
```

Comment:
What's the difference between this and `int`? Just curious -- haven't seen this before.